src/game/AI/ScriptDevAI/scripts/northrend/naxxramas/boss_sapphiron.cpp

```diff
@@ -16,339 +16,398 @@
+        if (m_instance)
+            m_instance->SetData(TYPE_SAPPHIRON, DONE);
     }
 
     void JustReachedHome() override
     {
-        if (m_pInstance)
-            m_pInstance->SetData(TYPE_SAPPHIRON, FAIL);
+        if (m_instance)
+            m_instance->SetData(TYPE_SAPPHIRON, FAIL);
     }
 
-    void JustSummoned(Creature* pSummoned) override
+    void JustSummoned(Creature* summoned) override
     {
-        if (pSummoned->GetEntry() == NPC_YOU_KNOW_WHO)
+        if (summoned->GetEntry() == NPC_BLIZZARD)
         {
             if (Unit* pEnemy = m_creature->SelectAttackingTarget(ATTACKING_TARGET_RANDOM, 0))
-                pSummoned->AI()->AttackStart(pEnemy);
+                summoned->AI()->AttackStart(pEnemy);
         }
     }
 
-    void MovementInform(uint32 uiType, uint32 /*uiPointId*/) override
+    void MovementInform(uint32 type, uint32 /*pointId*/) override
     {
-        if (uiType == POINT_MOTION_TYPE && m_Phase == PHASE_LIFT_OFF)
+        if (type == POINT_MOTION_TYPE && m_phase == PHASE_LIFT_OFF)
         {
-            DoScriptText(EMOTE_FLY, m_creature);
-            m_creature->HandleEmote(EMOTE_ONESHOT_LIFTOFF);
-            m_creature->SetLevitate(true);
-            m_Phase = PHASE_AIR_BOLTS;
+            // Summon the Wing Buffet NPC and cast the triggered aura to despawn it
+            if (DoCastSpellIfCan(m_creature, SPELL_SUMMON_WING_BUFFET) == CAST_OK)
+                DoCastSpellIfCan(m_creature, SPELL_DESPAWN_WING_BUFFET);
 
-            m_uiFrostBreathTimer = 5000;
-            m_uiIceboltTimer = 5000;
-            m_uiIceboltCount = 0;
+            // Actual take off
+            m_creature->HandleEmote(EMOTE_ONESHOT_LIFTOFF);
+            m_creature->SetHover(true);
+            m_creature->CastSpell(nullptr, SPELL_DRAGON_HOVER, TRIGGERED_OLD_TRIGGERED);
+            ResetCombatAction(SAPPHIRON_ICEBOLT, 8u * IN_MILLISECONDS);
         }
     }
 
-    void UpdateAI(const uint32 uiDiff) override
+    void HandleAirPhase()
     {
-        if (!m_creature->SelectHostileTarget() || !m_creature->GetVictim())
-            return;
-
-        switch (m_Phase)
+        if (m_creature->GetHealthPercent() > 10.f)
         {
-            case PHASE_GROUND:
-                if (m_uiCleaveTimer < uiDiff)
-                {
-                    if (DoCastSpellIfCan(m_creature->GetVictim(), SPELL_CLEAVE) == CAST_OK)
-                        m_uiCleaveTimer = urand(5000, 10000);
-                }
-                else
-                    m_uiCleaveTimer -= uiDiff;
-
-                if (m_uiTailSweepTimer < uiDiff)
-                {
-                    if (DoCastSpellIfCan(m_creature, m_bIsRegularMode ? SPELL_TAIL_SWEEP : SPELL_TAIL_SWEEP_H) == CAST_OK)
-                        m_uiTailSweepTimer = urand(7000, 10000);
-                }
-                else
-                    m_uiTailSweepTimer -= uiDiff;
+            m_phase = PHASE_LIFT_OFF;
+            for (uint32 action : groundPhaseActions)
```

Comment:
This is also wrong. Instead, used SetCombatScriptStatus(true);

This tells the AI that a script is ongoing and not to dispatch combat actions or spell lists. (obviously need to reset it on reset and on script end)

Reply:
Unless you want me to script the entire air phase with CustomActions this is the way to go

Reply:
It isnt. And I will bet my money. Illidan, onyxia and nightbane all work the same. They fly to their location, there they root themselves, end the combat script, and engage normal spell lists usage.

Reply:
And what I meant is that the disable is meant to be at the same time as when enabling the new ones. (cos thats when sapphiron swaps his spell lists)

Reply:
But then the timers would start at an unknowable point in time before he reaches the center of the room, leading to weird situations where he'd shoot his frostbolts while still walking towards the center

Reply:
I'd have to test if he can cast life drain or blizzard while walking to the center, but I think he doesn't...

Reply:
They wouldnt, thats what toggling SetCombatScriptStatus is for.